components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java

```diff
@@ -1690,6 +1748,71 @@ public static String mapDigestAlgorithm(Algorithm signatureAlgorithm) throws Ide
+     * @param encryptionAlgorithm JWT encryption algorithm
+     * @param spTenantDomain Service provider tenant domain
+     * @param clientId ID of the client
+     * @return encrypted JWT token
+     * @throws IdentityOAuth2Exception
+     */
+    public static JWT encryptJWT(JWTClaimsSet jwtClaimsSet, JWEAlgorithm encryptionAlgorithm, EncryptionMethod encryptionMethod,
+                                 String spTenantDomain, String clientId) throws IdentityOAuth2Exception{
+        if (JWEAlgorithm.RSA_OAEP.equals(encryptionAlgorithm) || JWEAlgorithm.RSA1_5.equals(encryptionAlgorithm) ||
+                JWEAlgorithm.RSA_OAEP_256.equals(encryptionAlgorithm)) {
+            return encryptWithRSA(jwtClaimsSet, encryptionAlgorithm, encryptionMethod, spTenantDomain, clientId);
+        } else {
+            throw new RuntimeException("Provided encryption algorithm: " + encryptionAlgorithm +
+                    " is not supported");
+        }
+    }
+
+    /**
+     * encrypt JWT token from RSA algorithm
+     *
+     * @param jwtClaimsSet contains JWT body
+     * @param encryptionAlgorithm JWT signing algorithm
+     * @param spTenantDomain Service provider tenant domain
+     * @param clientId ID of the client
+     * @return encrypted JWT token
+     * @throws IdentityOAuth2Exception
+     */
+    private static JWT encryptWithRSA(JWTClaimsSet jwtClaimsSet, JWEAlgorithm encryptionAlgorithm, EncryptionMethod encryptionMethod,
+                                      String spTenantDomain, String clientId) throws IdentityOAuth2Exception {
+        try {
+            if (StringUtils.isBlank(spTenantDomain)) {
+                spTenantDomain = MultitenantConstants.SUPER_TENANT_DOMAIN_NAME;
+                if (log.isDebugEnabled()) {
+                    log.debug("Assign super tenant domain as signing domain."); // TODO: 2/6/18 Refine
+                }
+            }
+
+            if (log.isDebugEnabled()) {
+                log.debug("Encrypting JWT using the algorithm: " + encryptionAlgorithm + " & key of the tenant: " +
+                        spTenantDomain);
+            }
+
+            int tenantId = IdentityTenantUtil.getTenantId(spTenantDomain);
+            Certificate publicCert = getX509CertOfOAuthApp(clientId, spTenantDomain);
+            Key publicKey = publicCert.getPublicKey();
+
+            JWEHeader header = new JWEHeader(encryptionAlgorithm, encryptionMethod);
+            header.setKeyID(getThumbPrint(spTenantDomain, tenantId));
+            header.setX509CertThumbprint(new Base64URL(getThumbPrint(spTenantDomain, tenantId)));
+
+            EncryptedJWT encryptedJWT = new EncryptedJWT(header, jwtClaimsSet);
+
+            JWEEncrypter encrypter = new RSAEncrypter((RSAPublicKey) publicKey);
+            encryptedJWT.encrypt(encrypter);
+
+            return encryptedJWT;
+        } catch (JOSEException e) {
+            throw new IdentityOAuth2Exception("Error occurred while encrypting JWT", e);
```

Comment:
Shall we add more context information like client_id, tenantDomain etc.?

Reply:
Sure